### PR TITLE
Merged my branch with quertenmont/74Xfwk branch

### PR DESCRIPTION
**Pull request by vischia**

### CODE REVIEW

**Comment by vischia** [comment]
Hey guys, 

The last commit from a few seconds ago 
( 1d839895d68571b849cad543cd9d46b75c815ecc )

contains a hack around the das_client problem.

Basically, das_client calls for the files list are now substituted by calls to the DBS API via https.

cURL, interfaced with the x509 proxy previously created, is used.

It works (all my jobs pick up an input file).

Cheers and good night!

Pietro

**Comment by vischia** [comment]
Commit d5a40fac8e7e90abcf45c3450951a9a5f5894b65 fixes the file list retrieval procedure.

I did not notice it yesterday because I was setting the X509_USER_PROXY environmental variable by hand at some point, but it was not set within the script when running cURL, resulting in no authentication.

The commit fixes this: I tested it submitting a couple times after having removed the existing proxy and logged out. It fully works now ;)

**Comment by vischia** [comment]
Looks like the additional "voms-proxy-init --noregen" was filling up the proxy file with additional certificate/proxies, and this at some point was surpassing some limits and the authentication could not take place.

Commit a759ec1be1c61392b285560e06628ac3863ea714 (included in this pull request) fixes this.

DISCLAIMER: It might be that for some submission systems other than lxplus the initialCommand parameter must contain the --noregen thing. I think it should not, but if you use s.th different from lxplus you should probably check. CRAB3 takes into account the issue itself, so it won't be a problem at all (and actually for crab tasks it should not contain the voms-proxy-init command at all)
Bottom line: it should be fine now: at least, in lxplus 100% of the jobs work fine :)

**Comment by vischia** [comment]
By the way, would everyone be fine if I remove the "split" parameter from the json file and define it (both for submission and plotting) by reading at runtime the number of files that constitute the sample (taken from dbs as the list)?

This would ensure that every sample is split in the maximum number of jobs possible, each job running on a single file. 

In principle, I think this would be optimal.

Does anyone have a different suggestion?

Cheers,
Pietro

P.S. is anyone receiving these comments to the pull request?

**Comment by hdelanno** [comment]
Hi Pietro,

Yes, at least I receive it.
Concerning the split parameter, I sometimes prefer to put it at 1. But this is really rare. I let Loic and you decide.

Cheers,

Hugo

Date: Sun, 28 Jun 2015 15:22:08 -0700

**Comment by vischia** [comment]
Hi Hugo,

thanks for the answer!

If the only exception is when it is put to 1, it might be dealt with in the
script in such a way that if it is 1 then it is left at 1, and if it is not
set, or set to any other value, it is taken dynamically from dbs.

If there are other exceptions, perhaps it is inefficient to code them all

Cheers,
Pietro

2015-06-29 1:02 GMT+02:00 hdelanno <notifications@github.com>:

> Hi Pietro,
>
> Yes, at least I receive it.
> Concerning the split parameter, I sometimes prefer to put it at 1. But
> this is really rare. I let Loic and you decide.
>
> Cheers,
>
> Hugo
>
> Date: Sun, 28 Jun 2015 15:22:08 -0700
> From: notifications@github.com
> To: 2l2v_fwk@noreply.github.com
> Subject: Re: [2l2v_fwk] Merged my branch with quertenmont/74Xfwk branch
> (#26)
>
> By the way, would everyone be fine if I remove the "split" parameter from
> the json file and define it (both for submission and plotting) by reading
> at runtime the number of files that constitute the sample (taken from dbs
> as the list)?
>
>
> This would ensure that every sample is split in the maximum number of jobs
> possible, each job running on a single file.
>
>
> In principle, I think this would be optimal.
>
>
> Does anyone have a different suggestion?
>
>
> Cheers,
>
> Pietro
>
>
> P.S. is anyone receiving these comments to the pull request?
>
>
> —
> Reply to this email directly or view it on GitHub.
>
>
>
>
>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/quertenmont/2l2v_fwk/pull/26#issuecomment-116358703>.
>



-- 
Pietro Vischia
LIP-Lisboa                   Av. Elias Garcia 14 - 1º, 1000-149 Lisboa,
Portugal
mobiles: 00351 96 724 9239 (PT), 0039 340 51 69 388 (IT)
office: 00351 21 799 5017
mail: pietro.vischia@gmail.com
skype: rambling_hadron

**Comment by quertenmont** [comment]
Hi Pietro,

Running one job per file is quite a bad idea...
Many samples have a lot (O(1000) files) and so you will create 
O(10x1000) output files that will have to be merged by the plotter at a 
later step.
The plotter step being a serial task, it's going to take much much 
longer and all the time you gain in having more jobs will be lost there.
Moreover, each of the O(10^4) output root files will contain histograms 
that will occupy a lot of space on disk and will cause quota issues on 
lxplus.
This might not be a problem for some analysis, but for the hZZ2l2v where 
we have many 2D histograms for signal reweighting it will quickly become 
an issue.

So removing the splitting might sounds like a good idea, but I don't 
think it is one.
What I would suggest you is therefore to either encore the "maximal 
splitting" option as 'split=-1' in your json  (and adapt the code in 
consequence)
OR even better, just add one possible argument to 
runAnalysisOverSample.py to specify that splitting argument should be 
replaced on the fly by the maximal possible splitting.
--> But you'll need to find a way to tell runPlotter what is the maximal 
splitting, because it does loop on all files between 0 and split.

Enjoy
Loic



Le 29/06/2015 01:10, Pietro Vischia a écrit :
> Hi Hugo,
>
> thanks for the answer!
>
> If the only exception is when it is put to 1, it might be dealt with 
> in the
> script in such a way that if it is 1 then it is left at 1, and if it 
> is not
> set, or set to any other value, it is taken dynamically from dbs.
>
> If there are other exceptions, perhaps it is inefficient to code them all
>
> Cheers,
> Pietro
>
> 2015-06-29 1:02 GMT+02:00 hdelanno <notifications@github.com>:
>
> > Hi Pietro,
> >
> > Yes, at least I receive it.
> > Concerning the split parameter, I sometimes prefer to put it at 1. But
> > this is really rare. I let Loic and you decide.
> >
> > Cheers,
> >
> > Hugo
> >
> > Date: Sun, 28 Jun 2015 15:22:08 -0700
> > From: notifications@github.com
> > To: 2l2v_fwk@noreply.github.com
> > Subject: Re: [2l2v_fwk] Merged my branch with quertenmont/74Xfwk branch
> > (#26)
> >
> > By the way, would everyone be fine if I remove the "split" parameter 
> from
> > the json file and define it (both for submission and plotting) by 
> reading
> > at runtime the number of files that constitute the sample (taken 
> from dbs
> > as the list)?
> >
> >
> > This would ensure that every sample is split in the maximum number 
> of jobs
> > possible, each job running on a single file.
> >
> >
> > In principle, I think this would be optimal.
> >
> >
> > Does anyone have a different suggestion?
> >
> >
> > Cheers,
> >
> > Pietro
> >
> >
> > P.S. is anyone receiving these comments to the pull request?
> >
> >
> > —
> > Reply to this email directly or view it on GitHub.
> >
> >
> >
> >
> >
> >
> > —
> > Reply to this email directly or view it on GitHub
> > 
> <https://github.com/quertenmont/2l2v_fwk/pull/26#issuecomment-116358703>.
> >
>
>
>
> -- 
> Pietro Vischia
> LIP-Lisboa Av. Elias Garcia 14 - 1º, 1000-149 Lisboa,
> Portugal
> mobiles: 00351 96 724 9239 (PT), 0039 340 51 69 388 (IT)
> office: 00351 21 799 5017
> mail: pietro.vischia@gmail.com
> skype: rambling_hadron
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/quertenmont/2l2v_fwk/pull/26#issuecomment-116359005>.
>


-- 
---------------------------------------------------------
Loic Quertenmont - European Organization for Nuclear Research (CERN) - CMS experiment
Office CERN   : bldg.40, RB-08
Phone CERN    : +41 2276 76966
Mobile BE     : +32 474 831531
WWW           : http://www.fynu.ucl.ac.be/users/l.quertenmont/

**Comment by quertenmont** [comment]
The change of the file TAGS.txt and README.txt are not correct.
Go back to the previous version

Also, I don't understand why you duplicated the runPLotter code and why you need so many json files.
We should avoid to duplicate things too much otherwise it will very quickly become messy.
Please check if you really need all of this.

Similarly, do not duplicate PatUtils::Id functions.
If you believe that one is better, just update the default one.

In general, I would prefer that you do push more often with specific changes, than having such huge PR modifying 32files in 71commits.  It makes it a bit hard to review.

Please apply the minimal fix described above before I merge the PR.
Loic

**Comment by vischia** [comment]
Hi Loic,

2015-06-29 11:54 GMT+02:00 Loic Quertenmont <notifications@github.com>:

> The change of the file TAGS.txt and README.txt are not correct.
> Go back to the previous version
>

Done!


> Also, I don't understand why you duplicated the runPLotter code and why
> you need so many json files.
> We should avoid to duplicate things too much otherwise it will very
> quickly become messy.
> Please check if you really need all of this.
>
I need that runPlotter because it is based on the previously called
"cutflow" histogram for the normalization, so I prefer to keep it.

On the json files you are right. Most of them remained there for inertia :D
I committed a cleanup of the json files I don't need anymore.

> Similarly, do not duplicate PatUtils::Id functions.
> If you believe that one is better, just update the default one.
>

You are right! I don't want to mess with the ID for other people, though
(perhaps personalized, whatever), so I had
duplicated the functions.

I did now update the base function, extending the possible IDs via the
enumerate. It is documented in the PatUtils.cc file.

In general, I would prefer that you do push more often with specific
> changes, than having such huge PR modifying 32files in 71commits. It makes
> it a bit hard to review.
>
Yes, I am sorry about that. I had pretty a quick-response-needed period in
which I was not able to take time to merge the new submission way with my
code, so I just kept pushing to my branch without pulling from your
repository.

I am gonna make pull requests fairly often from now on :)


Please apply the minimal fix described above before I merge the PR.
>

Everything is done and pushed ;)

Cheers,
Pietro


> Loic
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/quertenmont/2l2v_fwk/pull/26#issuecomment-116588535>.
>



-- 
Pietro Vischia
LIP-Lisboa                   Av. Elias Garcia 14 - 1º, 1000-149 Lisboa,
Portugal
mobiles: 00351 96 724 9239 (PT), 0039 340 51 69 388 (IT)
office: 00351 21 799 5017
mail: pietro.vischia@gmail.com
skype: rambling_hadron
